src/bp/nlu-core/intents/oos-intent-classfier.ts

```diff
@@ -0,0 +1,245 @@
+    trainInput: TrainInput,
+    noneIntent: Omit<Intent<Utterance>, 'contexts'>,
+    progress: (p: number) => void
+  ): Promise<string | undefined> {
+    const { allUtterances, nluSeed, intents } = trainInput
+    const { languageCode } = allUtterances[0]
+
+    const trainingOptions: MLToolkit.SVM.SVMOptions = {
+      c: [10], // so there's no grid search
+      kernel: 'LINEAR',
+      classifier: 'C_SVC',
+      seed: nluSeed
+    }
+
+    const noneUtts = noneIntent.utterances
+
+    if (!isPOSAvailable(languageCode) || noneUtts.length === 0) {
+      progress(1)
+      return
+    }
+
+    const vocab = this.getVocab(allUtterances)
+    const oos_points = featurizeOOSUtterances(noneUtts, vocab, this.tools)
+
+    const in_ctx_scope_points = _.chain(intents)
+      .filter(i => i.name !== NONE_INTENT)
+      .flatMap(i => featurizeInScopeUtterances(i.utterances, i.name))
+      .value()
+
+    const svm = new this.tools.mlToolkit.SVM.Trainer()
+
+    const model = await svm.train([...in_ctx_scope_points, ...oos_points], trainingOptions, progress)
+    return model
+  }
+
+  private async _trainInScopeSvm(
+    trainInput: TrainInput,
+    noneIntent: Omit<Intent<Utterance>, 'contexts'>,
+    progress: (p: number) => void
+  ): Promise<string> {
+    const baseIntentClf = new BaseIntentClassifier(this.tools)
+    const noneUtts = noneIntent.utterances.filter(u => u.tokens.filter(t => t.isWord).length >= 3)
+    const trainableIntents = trainInput.intents.filter(
+      i => i.name !== NONE_INTENT && i.utterances.length >= MIN_NB_UTTERANCES
+    )
+    const nAvgUtts = Math.ceil(_.meanBy(trainableIntents, i => i.utterances.length))
+
+    const lo = this.tools.seededLodashProvider.getSeededLodash()
+
+    const intents: Intent<Utterance>[] = [
+      ...trainInput.intents,
+      {
+        name: NONE_INTENT,
+        utterances: lo
+          .chain(noneUtts)
+          .shuffle()
+          .take(nAvgUtts * 2.5) // undescriptible magic n, no sens to extract constant
+          .value(),
+        contexts: [...trainInput.intents[0].contexts],
+        slot_definitions: []
+      }
+    ]
+
+    await baseIntentClf.train({ ...trainInput, intents }, progress)
+    return baseIntentClf.serialize()
+  }
+
+  private getVocab(utts: Utterance[]) {
+    return _.flatMap(utts, u => u.tokens.map(t => t.toString({ lowerCase: true })))
+  }
+
+  public serialize(): string {
+    if (!this.model) {
+      throw new Error('Intent classifier must be trained before calling serialize')
+    }
+    return JSON.stringify(this.model)
+  }
+
+  public load(serialized: string): void {
+    const model: Model = JSON.parse(serialized) // TODO: validate input
+
+    const { oosSvmModel, baseIntentClfModel, trainingVocab } = model
+
+    const baseIntentClf = new BaseIntentClassifier(this.tools)
+    baseIntentClf.load(baseIntentClfModel)
+
+    this.predictors = {
+      oosSvm: oosSvmModel ? new this.tools.mlToolkit.SVM.Predictor(oosSvmModel) : undefined,
```

Comment:
how's that possible ? oos is generated so will always be there no ?

Reply:
if `isPOSAvailable(languageCode)` is false then `oosSvmModel` is undefined.